packages/admin-ui/src/number-input/number-input.state.ts

```diff
@@ -0,0 +1,115 @@
+import { useEffect, useState } from 'react'
+import type { ButtonProps } from '../button'
+import {
+  isNumber,
+  isNumberAsString,
+  getIncrementedValue,
+  getDecrementedValue,
+} from './utils'
+
+export const useNumberInput = (props: NumberInputState) => {
+  const { initialValue, step, min, max, onChange } = props
+  const [currentValue, setCurrentValue] = useState(initialValue)
+
+  const handleChange: React.ChangeEventHandler<HTMLInputElement> = (event) => {
+    const { value } = event.target
+    const newValue = isNumberAsString(value) ? Number(value) : value
+
+    setCurrentValue(newValue)
+    onChange(newValue)
+  }
+
+  const handleIncrement: React.MouseEventHandler<HTMLButtonElement> = () => {
+    if (isNumber(currentValue)) {
+      const newValue = getIncrementedValue({
+        currentValue,
+        nextValue: currentValue + step,
+        min,
+        max,
+      })
+
+      setCurrentValue(newValue)
+      onChange(newValue)
+    } else {
+      // When the currentValue is a empty string and you click to increment
+      const newValue = max <= step ? max : step
+
+      setCurrentValue(newValue)
+      onChange(newValue)
+    }
+  }
+
+  const handleDecrement: React.MouseEventHandler<HTMLButtonElement> = () => {
+    if (isNumber(currentValue)) {
+      const newValue = getDecrementedValue({
+        currentValue,
+        nextValue: currentValue - step,
+        min,
+        max,
+      })
+
+      setCurrentValue(newValue)
+      onChange(newValue)
+    } else {
+      // When the currentValue is a empty string and you click to decrement
+      const newValue = min >= -step ? min : -step
+
+      setCurrentValue(newValue)
+      onChange(newValue)
+    }
+  }
```

Comment:
Just a suggestion and you can do the same for the increment
```suggestion
  const handleDecrement: React.MouseEventHandler<HTMLButtonElement> = (value: number | string) => {
    const currentValue = isNumber(value) ? value : 0
    const newValue = getDecrementedValue({
      currentValue,
      nextValue: currentValue - step,
      min,
      max,
    })

    setCurrentValue(newValue)
    onChange(newValue)
  }
```

Reply:
Using this suggestion it would be easier, but the behavior would be different from the native behavior of `input[type=number]` when there is no value. This is to avoid different behavior when it is used the button or the keyboard. What do you think about this?